nrf-hal-common/src/i2s.rs

```diff
@@ -0,0 +1,576 @@
+
+    /// Enables/disables I2S transmission (TX).
+    #[inline(always)]
+    pub fn tx_enabled(&self, enabled: bool) -> &Self {
+        self.i2s.config.txen.write(|w| w.txen().bit(enabled));
+        self
+    }
+
+    /// Enables/disables I2S reception (RX).
+    #[inline(always)]
+    pub fn rx_enabled(&self, enabled: bool) -> &Self {
+        self.i2s.config.rxen.write(|w| w.rxen().bit(enabled));
+        self
+    }
+
+    /// Sets MCK generator frequency.
+    #[inline(always)]
+    pub fn set_mck_frequency(&self, freq: MckFreq) -> &Self {
+        self.i2s
+            .config
+            .mckfreq
+            .write(|w| unsafe { w.mckfreq().bits(freq.into()) });
+        self
+    }
+
+    /// Sets MCK / LRCK ratio.
+    #[inline(always)]
+    pub fn set_ratio(&self, ratio: Ratio) -> &Self {
+        self.i2s
+            .config
+            .ratio
+            .write(|w| unsafe { w.ratio().bits(ratio.into()) });
+        self
+    }
+
+    /// Sets sample width.
+    #[inline(always)]
+    pub fn set_sample_width(&self, width: SampleWidth) -> &Self {
+        self.i2s
+            .config
+            .swidth
+            .write(|w| unsafe { w.swidth().bits(width.into()) });
+        self
+    }
+
+    /// Sets the sample alignment within a frame.
+    #[inline(always)]
+    pub fn set_align(&self, align: Align) -> &Self {
+        self.i2s.config.align.write(|w| w.align().bit(align.into()));
+        self
+    }
+
+    /// Sets the frame format.
+    #[inline(always)]
+    pub fn set_format(&self, format: Format) -> &Self {
+        self.i2s
+            .config
+            .format
+            .write(|w| w.format().bit(format.into()));
+        self
+    }
+
+    /// Sets the I2S channel configuation.
+    #[inline(always)]
+    pub fn set_channels(&self, channels: Channels) -> &Self {
+        self.i2s
+            .config
+            .channels
+            .write(|w| unsafe { w.channels().bits(channels.into()) });
+        self
+    }
+
+    /// Sets the transmit data buffer (TX).
+    #[inline(always)]
+    pub fn tx_buffer<B: I2SBuffer + ?Sized>(&self, buf: &B) -> Result<(), Error> {
+        if (buf.ptr() as usize) < SRAM_LOWER || (buf.ptr() as usize) > SRAM_UPPER {
+            return Err(Error::DMABufferNotInDataMemory);
+        }
+
+        if buf.maxcnt() > 16_384 {
+            return Err(Error::BufferTooLong);
+        }
+
+        self.i2s
+            .txd
+            .ptr
+            .write(|w| unsafe { w.ptr().bits(buf.ptr()) });
+        self.i2s
+            .rxtxd
+            .maxcnt
+            .write(|w| unsafe { w.bits(buf.maxcnt()) });
+
+        Ok(())
+    }
+
+    /// Sets the receive data buffer (RX).
+    #[inline(always)]
+    pub fn rx_buffer<B: I2SBuffer + ?Sized>(&self, buf: &mut B) -> Result<(), Error> {
+        if (buf.ptr() as usize) < SRAM_LOWER || (buf.ptr() as usize) > SRAM_UPPER {
+            return Err(Error::DMABufferNotInDataMemory);
+        }
+
+        if buf.maxcnt() > 16_384 {
```

Comment:
Can you move the size limit to a constant? Is it a hard limit of the I2S peripheral or is it just the `EASY_DMA_SIZE` divided by 4 (the size of the largest I2S data type)? Might be good to explain that in a comment.